### PR TITLE
fix: address regressions from v2 template refactor

### LINKED ISSUE #868: AKS-engine generated an invalid arm template
**Is this an ISSUE or FEATURE REQUEST?** (choose one):
---
Issue

**What version of aks-engine?**:
---
0.33.0

**Kubernetes version**:
1.13

**What happened**:
I generated an arm template using aks-engine, then tried to deploy it using the azure cli and received the following error: 

```
* null_resource.deploy-cluster: Error running command 'az group deployment create --name david-westus --resource-group david-westus --template-file ./_output/david-k8s-westus/azuredeploy.json --parameters ./_output/david-k8s-westus/azuredeploy.parameters.json': exit status 1. Output: ERROR: Azure Error: InvalidTemplate
Message: Deployment template validation failed: 'The template resource 'bcb77dc7-4265-57e9-ae28-2a61a5a610eb' at line '1' and column '58973' is not valid: Unable to parse language expression 'reference(concat('Microsoft.Compute/virtualMachineScaleSets/', variables('linuxpool1VMNamePrefix'), '2017-03-30', 'Full').identity.principalId': expected token 'RightParenthesis' and actual 'EndOfData'.. Please see https://aka.ms/arm-template-expressions for usage details.'.
```

**What you expected to happen**:
I expect the template pass az syntax validation, and deploy.

**How to reproduce it** (as minimally and precisely as possible):
@jackfrancis, please contact me directly and I will give you the repo which is generating the issue.

### PR DESCRIPTION
**Pull request by jackfrancis**
**Reason for Change**:

We have some identified regressions based on this api model input:

```
{
  "apiVersion": "vlabs",
  "location": "westus",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.13",
      "kubernetesConfig": {
        "enableRbac": true,
        "useManagedIdentity": true,
        "useInstanceMetadata": true,
        "cloudProviderBackoff": true,
        "cloudProviderBackoffRetries": 6,
        "cloudProviderBackoffJitter": 1,
        "cloudProviderBackoffDuration": 6,
        "cloudProviderBackoffExponent": 1.5,
        "cloudProviderRateLimit": false,
        "cloudProviderRateLimitQPS": 3,
        "cloudProviderRateLimitBucket": 10,
        "etcdDiskSizeGB": "4000",
        "networkPlugin": "azure",
        "enableEncryptionWithExternalKms": true,
        "keyVaultSku": "Premium",
        "controllerManagerConfig": {
          "--route-reconciliation-period": "20m"
        },
        "addons": [
          {
            "name": "kubernetes-dashboard",
            "enabled": false
          }
        ]
      }
    },
    "masterProfile": {
      "count": 3,
      "dnsPrefix": "",
      "vmSize": "Standard_D4_v3",
      "OSDiskSizeGB": 64,
      "storageProfile": "ManagedDisks",
      "availabilityZones": [
        "1",
        "2"
      ]
    },
    "agentPoolProfiles": [
      {
        "name": "linuxpool1",
        "count": 3,
        "vmSize": "Standard_D16_v3",
        "storageProfile": "ManagedDisks",
        "OSDiskSizeGB": 64,
        "availabilityProfile": "VirtualMachineScaleSets",
        "availabilityZones": [
          "1",
          "2"
        ]
      }
    ],
    "linuxProfile": {
      "adminUsername": "azureuser",
      "ssh": {
        "publicKeys": [
          {
            "keyData": ""
          }
        ]
      }
    }
  }
}
```

**Issue Fixed**:

Fixes #868

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jackfrancis** [comment]
@DavidParks8 FYI

File: pkg/engine/virtualmachinescalesets.go [hunk 1]
```diff
@@ -69,9 +69,7 @@ func CreateMasterVMSS(cs *api.ContainerService) VirtualMachineScaleSetARM {
 	}
 
 	if hasAvailabilityZones {
-		virtualMachine.Zones = &[]string{
-			"[parameters('availabilityZones')]",
-		}
+		virtualMachine.Zones = &masterProfile.AvailabilityZones
```
Comment (tariq1890) [inline pkg/engine/virtualmachinescalesets.go]: Why this change?

File: pkg/engine/virtualmachinescalesets.go [hunk 2]
```diff
@@ -374,9 +372,7 @@ func CreateAgentVMSS(cs *api.ContainerService, profile *api.AgentPoolProfile) Vi
 	}
 
 	if profile.HasAvailabilityZones() {
-		virtualMachineScaleSet.Zones = &[]string{
-			fmt.Sprintf("[parameters('%sAvailabilityZones')]", profile.Name),
-		}
+		virtualMachineScaleSet.Zones = &profile.AvailabilityZones
```
Comment (tariq1890) [inline pkg/engine/virtualmachinescalesets.go]: why this change?

File: pkg/engine/keyvaults.go [hunk 3]
```diff
@@ -61,7 +61,7 @@ func CreateKeyVaultVMAS(cs *api.ContainerService) map[string]interface{} {
 		} else {
 			for i := 0; i < masterCount; i++ {
 				accessPolicy := map[string]interface{}{
-					"objectId": fmt.Sprintf("[reference(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), %d'), '2017-03-30', 'Full').identity.principalId]", i),
+					"objectId": fmt.Sprintf("[reference(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), '%d'), '2017-03-30', 'Full').identity.principalId]", i),
```
Comment (tariq1890) [inline pkg/engine/keyvaults.go]: oh man :/

File: pkg/engine/keyvaults_test.go [hunk 4]
```diff
@@ -68,7 +68,7 @@ func TestCreateKeyVault(t *testing.T) {
 		"properties": map[string]interface{}{
 			"accessPolicies": []interface{}{
 				map[string]interface{}{
-					"objectId": "[reference(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), 0'), '2017-03-30', 'Full').identity.principalId]",
+					"objectId": "[reference(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), '0'), '2017-03-30', 'Full').identity.principalId]",
```
Comment (tariq1890) [inline pkg/engine/keyvaults_test.go]: * facepalm *

File: pkg/engine/virtualmachinescalesets.go (hunk 1 above)
Comment (jackfrancis) [reply pkg/engine/virtualmachinescalesets.go]: Because this is bad:

```
Azure Error: InvalidTemplate
Message: Deployment template parse failed: 'Error reading string. Unexpected token: StartArray. Path '[0]'.'.
```

File: pkg/engine/virtualmachinescalesets.go (hunk 2 above)
Comment (jackfrancis) [reply pkg/engine/virtualmachinescalesets.go]: ditto

**Comment by jackfrancis** [comment]
@tariq1890 feel free to start from scratch with the api model in the description. Basically, we just want to get that working again!

File: pkg/engine/virtualmachinescalesets.go (hunk 1 above)
Comment (jackfrancis) [reply pkg/engine/virtualmachinescalesets.go]: The real answer is the params evaluates to an array.  So it ends up being an array inside of an array (a pointer to), which isn't parseable.

**Comment by jackfrancis** [comment]
There's one last diff I noticed between generating the above api model against v0.32.3 and current master (basically pre- and post- template refactor switch):

Before:

```
{
      "apiVersion": "[variables('apiVersionAuthorizationSystem')]",
      "copy": {
        "count": "[variables('masterCount')]",
        "name": "vmLoopNode"
      },
      "name": "[guid(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), copyIndex(),'vmidentity'))]",
      "properties": {
        "principalId": "[reference(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), copyIndex()), '2017-03-30', 'Full').identity.principalId]",
        "roleDefinitionId": "[variables('contributorRoleDefinitionId')]"
      },
      "type": "Microsoft.Authorization/roleAssignments"
    },
```

After:

```
{
      "apiVersion": "[variables('apiVersionAuthorizationSystem')]",
      "copy": {
        "count": "[variables('masterCount')]",
        "name": "vmLoopNode"
      },
      "name": "[guid(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), copyIndex(), 'vmidentity'))]",
      "properties": {
        "principalId": "[reference(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), copyIndex()), '2017-03-30', 'Full').identity.principalId]",
        "roleDefinitionId": "[variables('readerRoleDefinitionId')]"
      },
      "type": "Microsoft.Authorization/roleAssignments"
    },
```

Note the changed `roleDefinitionId` value. Before the refactor, the resultant template would have a `"roleDefinitionId": "[variables('contributorRoleDefinitionId')]"` output. Now we're yielding `"roleDefinitionId": "[variables('readerRoleDefinitionId')]"`.

Again, the repro is just to use the above api model as template generation input before/after and ensure that any changes are purposeful.

File: pkg/engine/virtualmachinescalesets.go (hunk 1 above)
Comment (tariq1890) [reply pkg/engine/virtualmachinescalesets.go]: This change is good to go.

File: pkg/engine/virtualmachinescalesets.go (hunk 2 above)
Comment (tariq1890) [reply pkg/engine/virtualmachinescalesets.go]: Good to go
